server.go

```diff
@@ -1132,83 +1345,56 @@ func (sp *serverPeer) OnHeaders(_ *peer.Peer, msg *wire.MsgHeaders) {
 }
 
-// handleGetData is invoked when a peer receives a getdata wire message and is
-// used to deliver block and transaction information.
+// OnGetData is invoked when a peer receives a getdata wire message and is used
+// to deliver block and transaction information.
 func (sp *serverPeer) OnGetData(_ *peer.Peer, msg *wire.MsgGetData) {
 	// Ban peers sending empty getdata requests.
 	if len(msg.InvList) == 0 {
 		sp.server.BanPeer(sp)
 		return
 	}
 
-	numAdded := 0
-	notFound := wire.NewMsgNotFound()
-
-	length := len(msg.InvList)
 	// A decaying ban score increase is applied to prevent exhausting resources
 	// with unusually large inventory queries.
+	//
 	// Requesting more than the maximum inventory vector length within a short
-	// period of time yields a score above the default ban threshold. Sustained
+	// period of time yields a score above the default ban threshold.  Sustained
 	// bursts of small requests are not penalized as that would potentially ban
-	// peers performing IBD.
+	// peers performing the inintial chain sync.
+	//
 	// This incremental score decays each minute to half of its value.
-	if sp.addBanScore(0, uint32(length)*99/wire.MaxInvPerMsg, "getdata") {
+	numNewReqs := uint32(len(msg.InvList))
+	if sp.addBanScore(0, numNewReqs*99/wire.MaxInvPerMsg, "getdata") {
 		return
 	}
 
-	// We wait on this wait channel periodically to prevent queuing
-	// far more data than we can send in a reasonable time, wasting memory.
-	// The waiting occurs after the database fetch for the next one to
-	// provide a little pipelining.
-	var waitChan chan struct{}
-	doneChan := make(chan struct{}, 1)
-
-	for i, iv := range msg.InvList {
-		var c chan struct{}
-		// If this will be the last message we send.
-		if i == length-1 && len(notFound.InvList) == 0 {
-			c = doneChan
-		} else if (i+1)%3 == 0 {
-			// Buffered so as to not make the send goroutine block.
-			c = make(chan struct{}, 1)
-		}
-		var err error
-		switch iv.Type {
-		case wire.InvTypeTx:
-			err = sp.server.pushTxMsg(sp, &iv.Hash, c, waitChan)
-		case wire.InvTypeBlock:
-			err = sp.server.pushBlockMsg(sp, &iv.Hash, c, waitChan)
-		default:
-			peerLog.Warnf("Unknown type '%d' in inventory request from %s",
-				iv.Type, sp)
-			continue
-		}
-		if err != nil {
-			notFound.AddInvVect(iv)
-
-			// When there is a failure fetching the final entry
-			// and the done channel was sent in due to there
-			// being no outstanding not found inventory, consume
-			// it here because there is now not found inventory
-			// that will use the channel momentarily.
-			if i == len(msg.InvList)-1 && c != nil {
-				<-c
-			}
-		}
-		numAdded++
-		waitChan = c
+	// Prevent too many outstanding requests while still allowing the
+	// flexibility to send multiple simultaneous getdata requests that are
+	// served asynchronously.
+	numPendingGetDataReqs := len(sp.getDataQueue)
+	if numPendingGetDataReqs+1 > maxConcurrentGetDataReqs {
```

Comment:
Correct.

I originally was using a separate queue with a mutex and `sync.Cond` to signal updates where the mutex protected the length checks additional of items to the queue itself, but I decided to go with this channel-based approach instead because the code is much easier to read and, as you mentioned, it won't be an issue because I intentionally set the buffer sizes large enough.  They way far exceed anything a peer not trying to do malicious things will use in practice.

Moreover, in practice, the only way the limits would be hit without the peer being banned by the logic prior to this is if the peer is intentionally misbehaving and should be disconnected in that case anyway.

Reply:
the race could be avoided with a non-blocking write to the channel.  if this falls into the default case, then the queue was full.